src/libexpr/eval.hh

```diff
@@ -249,6 +249,45 @@ public:
     std::string_view forceString(Value & v, PathSet & context, const Pos & pos = noPos);
     std::string_view forceStringNoCtx(Value & v, const Pos & pos = noPos);
 
+    [[gnu::noinline, gnu::noreturn]]
+    void throwEvalError(const Pos & pos, const char * s) const;
+    [[gnu::noinline, gnu::noreturn]]
+    void throwTypeError(const Pos & pos, const char * s, const Value & v) const;
+    [[gnu::noinline, gnu::noreturn]]
+    void throwEvalError(const char * s, const std::string & s2) const;
+    [[gnu::noinline, gnu::noreturn]]
+    void throwEvalError(const Pos & pos, const Suggestions & suggestions, const char * s,
+        const std::string & s2) const;
+    [[gnu::noinline, gnu::noreturn]]
+    void throwEvalError(const Pos & pos, const char * s, const std::string & s2) const;
+    [[gnu::noinline, gnu::noreturn]]
+    void throwEvalError(const char * s, const std::string & s2, const std::string & s3) const;
+    [[gnu::noinline, gnu::noreturn]]
+    void throwEvalError(const Pos & pos, const char * s, const std::string & s2, const std::string & s3) const;
+    [[gnu::noinline, gnu::noreturn]]
+    void throwEvalError(const Pos & p1, const char * s, const Symbol & sym, const Pos & p2) const;
+    [[gnu::noinline, gnu::noreturn]]
+    void throwTypeError(const Pos & pos, const char * s) const;
+    [[gnu::noinline, gnu::noreturn]]
+    void throwTypeError(const Pos & pos, const char * s, const ExprLambda & fun, const Symbol & s2) const;
+    [[gnu::noinline, gnu::noreturn]]
+    void throwTypeError(const Pos & pos, const Suggestions & suggestions, const char * s,
+        const ExprLambda & fun, const Symbol & s2) const;
+    [[gnu::noinline, gnu::noreturn]]
+    void throwTypeError(const char * s, const Value & v) const;
+    [[gnu::noinline, gnu::noreturn]]
+    void throwAssertionError(const Pos & pos, const char * s, const std::string & s1) const;
+    [[gnu::noinline, gnu::noreturn]]
+    void throwUndefinedVarError(const Pos & pos, const char * s, const std::string & s1) const;
+    [[gnu::noinline, gnu::noreturn]]
+    void throwMissingArgumentError(const Pos & pos, const char * s, const std::string & s1) const;
+
+    [[gnu::noinline]]
+    void addErrorTrace(Error & e, const char * s, const std::string & s2) const;
+    [[gnu::noinline]]
+    void addErrorTrace(Error & e, const Pos & pos, const char * s, const std::string & s2) const;
```

Comment:
This is the big pain point for me. Now these hacky functions leak into a much used, public header. I understand the purpose of these functions, but I would really like to find a more DRY alternative to this.

Reply:
could also make them private and make all the *Expr classes friends of EvalState, but that feels *even more* of a hack than this. we might be able to make them `inline` instead and define them only in `eval-state.cpp` to make all other uses compile time errors, but that only adds to the hackiness as well :confused:

Reply:
How about going the other way around, and lookup the {Symbol,Position}Idx _before_ passing them to throw*Error ? I My understanding is that the stack usage should not change. If this machinery is indeed about reducing stack space, then it is about the same.

That would avoid having to expose all these dummy and ill-defined functions in the headers, and let them settle when we should find a way to get rid of them.

Obviously it would make the code a bit more verbose, due to the repeated lookups, but that is not a big deal for that code in particular. If we wanted it terse, we should first take care of the try/catch and re-throwing bits, that clutter the eval and primops files.

Reply:
we did that at first, but it does clutters the call sites something fierce. this was mostly a readability choice as well and not something we're particularly attached to or did for runtime metrics. the symbols table is a public member of EvalState though, so we *could* make that parameter explicit and remove the declarations from the header again. (which then creates another problem: where does the argument go? first, last, somewhere in between?)